crates/accelerate/src/commutation_analysis.rs

```diff
@@ -0,0 +1,197 @@
+            // index set containing the current gate
+            let commutation_entry = commutation_set
+                .entry(wire.clone())
+                .or_insert_with(|| vec![AIndexSet::from_iter([current_gate_idx])]);
+
+            // we can unwrap as we know the commutation entry has at least one element
+            let last = commutation_entry.last_mut().unwrap();
+
+            // if the current gate index is not in the set, check whether it commutes with
+            // the previous nodes -- if yes, add it to the commutation set
+            if !last.contains(&current_gate_idx) {
+                let mut all_commute = true;
+
+                for prev_gate_idx in last.iter() {
+                    // if the node is an input/output node, they do not commute, so we only
+                    // continue if the nodes are operation nodes
+                    if let (NodeType::Operation(packed_inst0), NodeType::Operation(packed_inst1)) =
+                        (&dag.dag[current_gate_idx], &dag.dag[*prev_gate_idx])
+                    {
+                        let op1 = packed_inst0.op.view();
+                        let op2 = packed_inst1.op.view();
+                        let params1 = packed_inst0.params_view();
+                        let params2 = packed_inst1.params_view();
+                        let qargs1 = dag.get_qargs(packed_inst0.qubits);
+                        let qargs2 = dag.get_qargs(packed_inst1.qubits);
+                        let cargs1 = dag.get_cargs(packed_inst0.clbits);
+                        let cargs2 = dag.get_cargs(packed_inst1.clbits);
+
+                        all_commute = commutation_checker.commute_inner(
+                            py,
+                            &op1,
+                            params1,
+                            packed_inst0.extra_attrs.as_deref(),
+                            qargs1,
+                            cargs1,
+                            &op2,
+                            params2,
+                            packed_inst1.extra_attrs.as_deref(),
+                            qargs2,
+                            cargs2,
+                            max_num_qubits,
+                        )?;
+                        if !all_commute {
+                            break;
+                        }
+                    } else {
+                        all_commute = false;
+                        break;
+                    }
+                }
+
+                if all_commute {
+                    // all commute, add to current list
+                    last.insert(current_gate_idx);
+                } else {
+                    // does not commute, create new list
+                    commutation_entry.push(AIndexSet::from_iter([current_gate_idx]))
+                }
+            }
+
+            node_indices.insert(
+                (current_gate_idx, wire.clone()),
+                commutation_entry.len() - 1,
+            );
+        }
+    }
```

Comment:
This is a probably a good candidate for parallelization. We can naively make this a parallel iterator for the outer loop like:

```rust
0..dag.num_qubits().into_par_iter().map(|qubit| Wire::Qubit(Qubit(qubit as u32))).for_each(|wire| {
```

We'll probably have to play some games around synchronization, although we can probably change it into an iterator collector. But we should do this in a follow up because it might require some deeper changes (as the compiler points them out).

Reply:
Yeah good point -- I'll have a look in parallel 🙂 I was trying to propagate the Result coming from `commute_inner` but if we collect then we can still do that 👍🏻